File: Motocitizen/src/motocitizen/fragments/DetailMessagesFragment.java
package motocitizen.fragments;

import android.os.Bundle;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.PopupWindow;
import android.widget.ScrollView;

import org.jetbrains.annotations.NotNull;
import org.json.JSONException;
import org.json.JSONObject;

import motocitizen.activity.AccidentDetailsActivity;
import motocitizen.content.NewContent;
import motocitizen.content.accident.Accident;
import motocitizen.content.message.Message;
import motocitizen.database.StoreMessages;
import motocitizen.dictionary.Content;
import motocitizen.main.R;
import motocitizen.network.ApiRequest;
import motocitizen.network.requests.SendMessageRequest;
import motocitizen.user.User;
import motocitizen.utils.ToastUtils;
import motocitizen.utils.popups.MessagesPopup;

public class DetailMessagesFragment extends AccidentDetailsFragments {

    private ImageButton newMessageButton;

    private ScrollView activityDetailsMessagesScroll;
    private EditText   mcNewMessageText;
    private View       newMessageArea;
    private ViewGroup  messagesTable;

    public DetailMessagesFragment() {
    }

    public static DetailMessagesFragment newInstance(int accID, String userName) {
        DetailMessagesFragment fragment = new DetailMessagesFragment();
        Bundle                 args     = new Bundle();
        args.putInt(ACCIDENT_ID, accID);
        args.putString(USER_NAME, userName);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, final ViewGroup container, Bundle savedInstanceState) {
        View viewMain = inflater.inflate(R.layout.fragment_detail_messages, container, false);

        newMessageButton = (ImageButton) viewMain.findViewById(R.id.new_message_send);
        mcNewMessageText = (EditText) viewMain.findViewById(R.id.new_message_text);
        messagesTable = (ViewGroup) viewMain.findViewById(R.id.details_messages_table);
        newMessageArea = viewMain.findViewById(R.id.new_message_area);

        activityDetailsMessagesScroll = (ScrollView) viewMain.findViewById(R.id.activity__details_messages_scroll);

        //mcNewMessageText.addTextChangedListener(new NewMessageTextWatcher());
        newMessageButton.setOnClickListener(v -> {
            String temp = mcNewMessageText.getText().toString().replaceAll("\\s", "");
            if (temp.length() > 0) {
                new SendMessageRequest(mcNewMessageText.getText().toString(), accidentID, new SendMessageCallback());
                mcNewMessageText.setText("");

            }
        });
        update();

        return viewMain;
    }

    /**
     * Обновление сообщений в списке
     */
    private void update() {
        messagesTable.removeAllViews();
        final Accident accident = ((AccidentDetailsActivity) getActivity()).getCurrentPoint();

        Integer preLast = 0;
        Integer last    = 0;
        for (Integer key : accident.getMessages().keySet()) {
            if (last != 0) processMessage(accident, last, preLast, key);
            preLast = last;
            last = key;
        }
        if (accident.getMessages().size() > 0) {
            updateUnreadMessages(accident.getId(), last);
            processMessage(accident, last, preLast, 0);
        }
        setupAccess();
    }

    private void processMessage(Accident accident, Integer current, Integer last, Integer next) {
        Message message = accident.getMessages().get(current);
        View    row     = message.getRow(getActivity(), last, next);
        row.setOnLongClickListener(new MessageRowLongClickListener(message, accident));
        messagesTable.addView(row);
    }

    private void updateUnreadMessages(int accidentId, int messageId) {
        StoreMessages.setLast(accidentId, messageId);
    }

    private void setupAccess() {
        newMessageArea.setVisibility(User.getInstance(getActivity()).isStandard() ? View.VISIBLE : View.INVISIBLE);
    }

    private class SendMessageCallback implements ApiRequest.RequestResultCallback {
        @Override
        public void call(@NotNull JSONObject result) {
            if (result.has("error")) {
                getActivity().runOnUiThread(() -> {
                    try {
                        ToastUtils.show(getActivity(), result.getString("error"));
                    } catch (JSONException e) {
                        ToastUtils.show(getActivity(), "Неизвестная ошибка" + result.toString());
                        e.printStackTrace();
                    }
                });
            }
            NewContent.INSTANCE.requestUpdate(response -> getActivity().runOnUiThread(() -> {
//            Content.getInstance().requestUpdate(response -> getActivity().runOnUiThread(() -> {
                if (!response.has("error")) {
                    NewContent.INSTANCE.parseJSON(response, accidentID);
//                    Content.getInstance().parseJSON(response, accidentID);
                }
                ((AccidentDetailsActivity) getActivity()).update();
                update();
                activityDetailsMessagesScroll.post(() -> activityDetailsMessagesScroll.fullScroll(ScrollView.FOCUS_DOWN));
            }));
        }
    }

    private class MessageRowLongClickListener implements View.OnLongClickListener {
        private final Message  message;
        private final Accident accident;

        MessageRowLongClickListener(Message message, Accident accident) {
            this.message = message;
            this.accident = accident;
        }

        @Override
        public boolean onLongClick(View v) {
            PopupWindow popupWindow;
            popupWindow = (new MessagesPopup(getActivity(), message.getId(), accident.getId())).getPopupWindow(getActivity());
            int viewLocation[] = new int[ 2 ];
            v.getLocationOnScreen(viewLocation);
            popupWindow.showAtLocation(v, Gravity.NO_GRAVITY, viewLocation[ 0 ], viewLocation[ 1 ]);
            return true;
        }
    }
}

File: Motocitizen/src/motocitizen/fragments/DetailMessagesFragment.kt
package motocitizen.fragments

import android.app.Fragment
import android.os.Bundle
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import android.widget.PopupWindow
import android.widget.ScrollView
import motocitizen.activity.AccidentDetailsActivity
import motocitizen.content.Content
import motocitizen.content.accident.Accident
import motocitizen.content.message.Message
import motocitizen.database.StoreMessages
import motocitizen.main.R
import motocitizen.network.CoreRequest
import motocitizen.network.requests.SendMessageRequest
import motocitizen.rows.message.MessageRowFactory
import motocitizen.user.User
import motocitizen.utils.popups.MessagesPopup
import motocitizen.utils.show
import org.json.JSONException
import org.json.JSONObject
import java.util.*

class DetailMessagesFragment() : Fragment() {
    private val ROOT_LAYOUT = R.layout.fragment_detail_messages
    private val MESSAGES_VIEW = R.id.details_messages_table
    private val FORM_VIEW = R.id.new_message_area
    private val INPUT_FIELD = R.id.new_message_text
    private val SEND_BUTTON = R.id.new_message_send
    private val SCROLL_VIEW = R.id.activity__details_messages_scroll

    private lateinit var rootView: View
    private lateinit var scrollView: ScrollView
    private lateinit var messagesView: ViewGroup
    private lateinit var accident: Accident

    constructor(accident: Accident) : this() {
        this.accident = accident
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        rootView = inflater.inflate(ROOT_LAYOUT, container, false)
        messagesView = rootView.findViewById(MESSAGES_VIEW) as ViewGroup
        scrollView = rootView.findViewById(SCROLL_VIEW) as ScrollView

        setupNewMessageForm()

        update()//todo exterminatus

        return rootView
    }

    private fun setupNewMessageForm() {
        val formView = rootView.findViewById(FORM_VIEW)

        if (!User.isStandard) {
            formView.visibility = View.INVISIBLE
            return
        }
        formView.visibility = View.VISIBLE

        val inputField = rootView.findViewById(INPUT_FIELD) as EditText

        rootView.findViewById(SEND_BUTTON).setOnClickListener { _ ->
            val text = inputField.text.toString().replace("\\s".toRegex(), "")
            if (text.isNotEmpty()) {
                SendMessageRequest(inputField.text.toString(), accident.id, SendMessageCallback())
                inputField.setText("")
            }
        }
    }

    /**
     * Обновление сообщений в списке
     */
    private fun update() {
        messagesView.removeAllViews()

        if (accident.messages.isEmpty()) return

        var last = 0
        val group = ArrayList<Message>()
        accident.messages.forEach { message ->
            if (last != message.owner && last != 0) {
                addMessageRows(group)
                group.clear()
            }
            group.add(message)
            last = message.owner
        }
        addMessageRows(group)

        updateUnreadMessages(accident.id, last)
    }

    private fun addMessageRows(list: List<Message>) {
        if (list.size == 1) {
            drawRow(MessageRowFactory.makeOne(activity, list.first()), list.first())
        } else {
            drawRow(MessageRowFactory.makeFirst(activity, list.first()), list.first())
            (1..list.size - 2).forEach { drawRow(MessageRowFactory.makeMiddle(activity, list[it]), list[it]) }
            drawRow(MessageRowFactory.makeLast(activity, list.last()), list.last())
        }
    }

    private fun drawRow(view: View, message: Message) {
        view.setOnLongClickListener(MessageRowLongClickListener(message))
        messagesView.addView(view)
    }

    private fun updateUnreadMessages(accidentId: Int, messageId: Int) {
        StoreMessages.setLast(accidentId, messageId)
    }

    private inner class SendMessageCallback : CoreRequest.RequestResultCallback {
        override fun call(response: JSONObject) {
            try {
                if (response.getJSONObject("e").has("c")) {
                    val text = response.getJSONObject("e").getString("t")
                    activity.runOnUiThread { show(activity, text) }
                }
            } catch (e: JSONException) {
                show(activity, "Неизвестная ошибка" + response.toString())
                e.printStackTrace()
            }

            Content.requestDetailsForAccident(accident, object : CoreRequest.RequestResultCallback {
                override fun call(response: JSONObject) {
                    activity.runOnUiThread {
                        (activity as AccidentDetailsActivity).update()
                        update()
                        scrollView.post { scrollView.fullScroll(ScrollView.FOCUS_DOWN) }
                    }
                }
            })
        }
    }

    private inner class MessageRowLongClickListener internal constructor(private val message: Message) : View.OnLongClickListener {

        override fun onLongClick(view: View): Boolean {
            val popupWindow: PopupWindow = MessagesPopup(activity, message.id, accident.id).getPopupWindow(activity)
            val viewLocation = IntArray(2)
            view.getLocationOnScreen(viewLocation)
            popupWindow.showAtLocation(view, Gravity.NO_GRAVITY, viewLocation[0], viewLocation[1])
            return true
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        outState.putInt(AccidentDetailsActivity.ACCIDENT_ID_KEY, accident.id)
        super.onSaveInstanceState(outState)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        if (savedInstanceState == null) return
        accident = Content.accidents[savedInstanceState.getInt(AccidentDetailsActivity.ACCIDENT_ID_KEY)]!!
    }
}
